### ISSUE TITLE
Zip File not with many files on Mac gets FileDoesNotExistException because of open files

### ISSUE THREAD

**Issue by SabineMa** [guillep/pharo-core#7960]
**Describe the bug**
when I create a zip file with a lot of files, then I get a FileDoesNotExistException. 

**To Reproduce**
```
| theWorkingdir theNumberOfFiles theZip |
    theWorkingdir := FileSystem workingDirectory.
    theNumberOfFiles := 300.
    (theWorkingdir / 'testfiles') ensureCreateDirectory.
    1 to: theNumberOfFiles do: [ :i | 
        (theWorkingdir / ('testfiles/test-' , i asString))
            writeStreamDo: [ :stream | stream nextPutAll: 'file contents' ];
            yourself ].
    theZip := ZipArchive new.
    (theWorkingdir / 'testfiles') asFileReference files
        do: [ :eachFile | theZip addFile: eachFile ].
    theZip writeToFile: theWorkingdir / 'testfiles/result.zip'.
    theZip close
```
produces a walkback on Mac with pahro 8/9
**my launchctl limit maxfiles is 256**

I think this is related to that:

 
 
![Bildschirmfoto 2020-12-03 um 14 04 50](https://user-images.githubusercontent.com/13763544/101049227-9e6f1500-3583-11eb-9684-eb1f86dbef5d.png)

-> there are all files open while adding them to the zip. Perhaps it is in `ZipNewFileMember>>from:`

```
Virtual Machine
---------------
/Users/sabine/Documents/Pharo/vms/80-x64/Pharo.app/Contents/MacOS/Pharo
CoInterpreter VMMaker.oscog-eem.2509 uuid: 91e81f64-95de-4914-a960-8f842be3a194 Feb  7 2019
StackToRegisterMappingCogit VMMaker.oscog-eem.2509 uuid: 91e81f64-95de-4914-a960-8f842be3a194 Feb  7 2019
VM: 201902062351Date: Wed Feb 6 15:51:18 2019 CommitHash: a838346 Plugins: 201902062351 

Mac OS X built on Feb  7 2019 00:01:47 UTC Compiler: 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)
VMMaker versionString VM: 201902062351 Date: Wed Feb 6 15:51:18 2019 CommitHash: a838346 Plugins: 201902062351 
CoInterpreter VMMaker.oscog-eem.2509 uuid: 91e81f64-95de-4914-a960-8f842be3a194 Feb  7 2019
StackToRegisterMappingCogit VMMaker.oscog-eem.2509 uuid: 91e81f64-95de-4914-a960-8f842be3a194 Feb  7 2019
```

**Expected behavior**
I am creating bookkeeping reports which have to have a clear defined format and that results in zip files with a lot of (small) files

**Additional context**
When I close the zip file and open it again like this then the file is created and no open files left.
just for info, this is not a solution.

```
| theWorkingdir theNumberOfFiles theZip theZipFilereference theFiles |
    theWorkingdir := FileSystem workingDirectory.
    theNumberOfFiles := 300.
    (theWorkingdir / 'testfiles') exists
        ifTrue: [ (theWorkingdir / 'testfiles') deleteAll ].
    (theWorkingdir / 'testfiles') ensureCreateDirectory.
    theZipFilereference := theWorkingdir / 'testfiles/result.zip'.
    1 to: theNumberOfFiles do: [ :i | 
        (theWorkingdir / ('testfiles/test-' , i asString))
            writeStreamDo: [ :stream | stream nextPutAll: 'file contents' ];
            yourself ].
    theZip := ZipArchive new.
    theFiles := (theWorkingdir / 'testfiles') asFileReference files.
    theFiles from: 1 to: 99 do: [ :eachFile | theZip addFile: eachFile ].
    theZip writeToFile: theZipFilereference.
    theZip close.
    theZip := ZipArchive new readFrom: theZipFilereference.
    theFiles
        from: 100
        to: 199
        do: [ :eachFile | theZip addFile: eachFile ].
     theZip close.
    theZip := ZipArchive new readFrom: theZipFilereference.
    theFiles
        from: 200
        to: 300
        do: [ :eachFile | theZip addFile: eachFile ].
    theZip close
```
increasing the limit of open files on my mac
`sudo launchctl limit maxfiles 1000 unlimited`
allows me to workaround this

**Comment by seandenigris** [comment]
I've encountered this quite a bit in images launched via Launcher (not sure if that detail is relevant)

**Comment by guillep** [comment]
Looks like defining the following solves that issue:

```smalltalk
ZipNewFileMember >> endRead
	
	super endRead.
	self close
```

Could somebody check with a realistic example?

**Comment by SabineMa** [comment]
`ZipNewFileMember >> endRead `  is not called before the FileDoesNotExistException. 

`ZipNewFileMember >> endRead` is called when writing the zip. This is too late because the FileDoesNotExistException error occurs when adding the files.

I tried to close the file in `>>addFile: `like this:

```
addFile: aFileReference
    ^(self addFile: aFileReference as: aFileReference path pathString) close
```

But then I get a primitive error when writing the file  after >>rewindData.
Perhaps that could be a way to a solution?

**Comment by guillep** [comment]
Yes, sorry, I was apparently dead the other day :). I've just come back to it and checked my code.

To make it short, the problem is that files are opened eagerly when added to the Zip archive. Even before they are processed. So if you try to add X files, the Zip archive will **OPEN** X files. But the system gives us a limit of Y files, so BOOM.

In my insomnia solution of the other day, in addition of closing the file on `endRead`, what I did is that file streams are open lazily on their first read.

Here are the interesting code snippets to test with:
```smalltalk
ZipNewFileMember >> from: aFileReference
	| entry |
	"Now get the size, attributes, and timestamps, and see if the file exists"
	self localFileName: (externalFileName := aFileReference path pathString).
	entry := aFileReference entry.
	compressedSize := uncompressedSize := entry size.
	desiredCompressionMethod := compressedSize > 0 ifTrue: [ CompressionDeflated ] ifFalse: [ CompressionStored ].
	self flag: 'When we replace Files with FileSystem, the following line won''t have to jump throught hoops (FS returns aDateAndTime)'.
	self modifiedAt: entry modificationTime.

```

```smalltalk
ZipNewFileMember >> readRawChunk: n
	stream ifNil: [ stream := externalFileName asFileReference binaryReadStream ].
	^stream next: n
```

```smalltalk
ZipNewFileMember >> endRead
	super endRead.
	self close
```

I've tested the example and it works. But it is not clear to me if other classes in the hierarchy (like ZipFileArchive) require something similar :/

I'll try to write some tests for it.

**Comment by SabineMa** [comment]
Hi @guillep,

thank you for your time.

What I did for checking your fix is changing the 2 methods` from:` and `readRawChunk:` and adding the `endRead` in my current Pharo8 dev image. And I deleted the rewindData method. I changed my maxfiles back to 250 with 
sudo launchctl limit maxfiles 250 unlimited`
`
Then my small example from above works and also creating a zip file from my dev with more than 700 smalll files and that worked, too.

I also watched the open files in my macs activity monitor and there were not all files open at one time (as before).
